pandas/core/ops/array_ops.py

```diff
@@ -132,3 +153,232 @@ def na_arithmetic_op(left, right, op, str_rep, eval_kwargs):
         result = masked_arith_op(left, right, op)
 
     return missing.dispatch_fill_zeros(op, left, right, result)
+
+
+def arithmetic_op(
+    left: Union[np.ndarray, ABCExtensionArray],
+    right: Any,
+    op,
+    str_rep: str,
+    eval_kwargs: Dict[str, str],
+):
+    """
+    Evaluate an arithmetic operation `+`, `-`, `*`, `/`, `//`, `%`, `**`, ...
+
+    Parameters
+    ----------
+    left : np.ndarray or ExtensionArray
+    right : object
+        Cannot be a DataFrame or Index.  Series is *not* excluded.
+    op : {operator.add, operator.sub, ...}
+        Or one of the reversed variants from roperator.
+
+    Returns
+    -------
+    ndarrray or ExtensionArray
+        Or a 2-tuple of these in the case of divmod or rdivmod.
+    """
+
+    from pandas.core.ops import (
+        maybe_upcast_for_op,
+        should_extension_dispatch,
+        dispatch_to_extension_op,
+    )
+
+    keep_null_freq = isinstance(
+        right,
+        (
+            ABCDatetimeIndex,
+            ABCDatetimeArray,
+            ABCTimedeltaIndex,
+            ABCTimedeltaArray,
+            Timestamp,
+        ),
+    )
+
+    # NB: We assume that extract_array has already been called on `left`, but
+    #  cannot make the same assumption about `right`.  This is because we need
+    #  to define `keep_null_freq` before calling extract_array on it.
+    lvalues = left
+    rvalues = extract_array(right, extract_numpy=True)
+
+    rvalues = maybe_upcast_for_op(rvalues, lvalues.shape)
+
+    if should_extension_dispatch(left, rvalues) or isinstance(
+        rvalues, (ABCTimedeltaArray, ABCDatetimeArray, Timestamp)
+    ):
+        # TimedeltaArray, DatetimeArray, and Timestamp are included here
+        #  because they have `freq` attribute which is handled correctly
+        #  by dispatch_to_extension_op.
+        res_values = dispatch_to_extension_op(op, lvalues, rvalues, keep_null_freq)
+
+    else:
+        with np.errstate(all="ignore"):
+            res_values = na_arithmetic_op(lvalues, rvalues, op, str_rep, eval_kwargs)
+
+    return res_values
+
+
+def comparison_op(
+    left: Union[np.ndarray, ABCExtensionArray], right: Any, op
+) -> Union[np.ndarray, ABCExtensionArray]:
+    """
+    Evaluate a comparison operation `=`, `!=`, `>=`, `>`, `<=`, or `<`.
+
+    Parameters
+    ----------
+    left : np.ndarray or ExtensionArray
+    right : object
+        Cannot be a DataFrame, Series, or Index.
```

Comment:
Is it possible to be more precise and say that it is a ndarray, ExtensionArray or scalar? 
Or are there still other types of objects that can be passed?

Reply:
any wacky thing that you can pass to `Series.__add__` can end up here with the exception of those three classes.

Reply:
(maybe `__eq__` would be a better example)

Reply:
But eg a list, is that already coerced to an array by the time it is here? 
Depending on what work has been done before this, the  "any wacky thing you can pass" can still be considered as "a scalar"

Reply:
Some coercion (including list) is done a few lines down on L245-248.

Reply:
But the (long term, not here) idea is that those checks are also moved to helper functions/decorators, so that they can be applied consistently, and so you can know here in this function you have either a ndarray/EA or scalar?

Reply:
Yes to everything up through "applied consistently".  For the "or scalar" I'm reticent to make a definitive statement because we can brainstorm scalar-like things that wouldn't satisfy `lib.is_scalar` (possible also some special cases for tuple)

Reply:
> scalar-like things that wouldn't satisfy lib.is_scalar

That's only a problem for object / generic EAs (ones we don't know) right? I mean for all other types you could have that guarantee at this location in the code?

Reply:
No.  A user can pass something like `ser + type(ser)` or `ser + pd` and we an get here.

Reply:
But for all built-in types except object dtype, we know that such objects are invalid, so that could (consistently) be checked before getting here? (similarly like you want to move the length check out I thought)

(again, about long term plan, not this PR)

Reply:
I guess we could in principle.  I don't see it as a priority.